tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -6018,21 +6019,25 @@ void CGMSHLSLRuntime::EmitHLSLOutParamConversionInit(
         // copy to let the lower not happen on argument when calle is noinline
         // or extern functions. Will do it in HLLegalizeParameter after known
         // which functions are extern but before inline.
-        bool bConstGlobal = false;
         Value *Ptr = argAddr;
         while (GEPOperator *GEP = dyn_cast_or_null<GEPOperator>(Ptr)) {
           Ptr = GEP->getPointerOperand();
         }
-        if (GlobalVariable *GV = dyn_cast_or_null<GlobalVariable>(Ptr)) {
-          bConstGlobal = m_ConstVarAnnotationMap.count(GV) | GV->isConstant();
-        }
         // Skip copy-in copy-out when safe.
         // The unsafe case will be global variable alias with parameter.
         // Then global variable is updated in the function, the parameter will
         // be updated silently. For non global variable or constant global
         // variable, it should be safe.
-        if (argAddr &&
-            (isa<AllocaInst>(Ptr) || isa<Argument>(Ptr) || bConstGlobal)) {
```

Comment:
Why const global is not safe?

Reply:
The problem is that we don't capture the const-ness correctly in the source. If you look at the `gep_arg_nocopy.hlsl` test, if we don't copy the array in at the first call it is writable back to the global.

We could omit copying const globals in cases where the source appropriately annotates them as const, but promoting what should be a const global to a non-const pointer is incorrect code generation.

Reply:
Worth noting there is a larger semantic problem that since we don't capture the constness in the AST we don't respect it with reasonable source diagnostics.

In [this example](https://godbolt.org/z/cc8En7Eh8) based on the `gep_arg_nocopy.hlsl` test, we get the right behavior in the end, but I think we only get there by doing an unsafe transformation. If you look at the `-fcgl` IR generation we're generating a store to what should be a constant global. I assume at some point in our legalization passes we drop that store.

Reply:
We do capture constant variable inside constant buffers.
Those should be safe.